kernel/component-manager/src/main/java/org/sakaiproject/component/impl/ComponentManagerShim.java

```diff
@@ -102,14 +107,24 @@ public <T> T get(Class<T> iface) {
             return applicationContext.getBean(iface);
         } catch (NoUniqueBeanDefinitionException e) {
             if (lenientLookup) {
-                log.warn("No unique bean found for class: {} -- lenient retrieval enabled, retrying by name", iface.getName());
+                if ( uniqueFailures.contains(iface.getName()) ) {
+                    log.debug("No unique bean found for class: {} -- lenient retrieval enabled, retrying by name", iface.getName());
+                } else {
+                    log.warn("No unique bean found for class: {} -- lenient retrieval enabled, retrying by name - further messages will be log.debug", iface.getName());
+                    uniqueFailures.add(iface.getName());
+                }
```

Comment:
This adds enough nonfunctional bulk to these methods that I would like to see the logging logic extracted. The shape is the same for all of them, so one method would do. Something like this, maybe?

```java
logLenientLookup("No unique bean found for class: {} -- lenient retrieval enabled, retrying by name", iface.getName(), nonuniqueFailures);

...

private logLenientLookup(String msg, String name, Set<String> alreadyLogged) {
    if (alreadyLogged.contains(name)) {
        log.debug(msg, name);
    } else {
        log.warn(msg + " (further messages will be log.debug)", name);
        alreadyLogged.add(name);
    }
}
```

Reply:
This will be in the next push.  https://www.youtube.com/watch?v=85RZMIAL7vM :)